translate/tools/pocount.py

```diff
@@ -116,12 +189,38 @@ def calcstats_old(filename):
         + stats["fuzzysourcewords"]
         + stats["untranslatedsourcewords"]
     )
+
+    stats["extended"] = file_extended_totals(units, wordcounts)
+
     return stats
 
 
-def calcstats(filename):
-    statscache = statsdb.StatsCache()
-    return statscache.filetotals(filename, extended=True)
+def file_extended_totals(units, wordcounts):
+    """
+        Provide extended statuses (used by XLIFF)
+    """
+
+    stats = dict()
+
+    for unit in units:
+        state = unit.get_state_n()
+        # print(state, extended_state_strings[state])
+        state_stats = stats.get(extended_state_strings[state], {})
+        unit_count = state_stats.get("units", 0)
+        sourcewords_count = state_stats.get("sourcewords", 0)
+        targetwords_count = state_stats.get("targetwords", 0)
+
+        unit_count += 1
+        sourcewords_count += wordcounts[id(unit)][0]
+        targetwords_count += wordcounts[id(unit)][1]
+
+        state_stats["units"] = unit_count
+        state_stats["sourcewords"] = sourcewords_count
+        state_stats["targetwords"] = targetwords_count
```

Comment:
This all could be heavily simplified by using defaultdict (the patch would need importing it as well):

```suggestion
        state_stats = stats.get(extended_state_strings[state], defaultdict(int))

        state_stats["units"] += 1
        state_stats["sourcewords"] += wordcounts[id(unit)][0]
        state_stats["targetwords"] += wordcounts[id(unit)][1]
```